go/jsonclient/client_test.go

```diff
@@ -0,0 +1,304 @@
+package jsonclient
+
+import (
+	"encoding/json"
+	"fmt"
+	"math"
+	"net/http"
+	"net/http/httptest"
+	"reflect"
+	"strconv"
+	"strings"
+	"testing"
+	"time"
+
+	"golang.org/x/net/context"
+
+	"github.com/google/certificate-transparency/go/testdata"
+)
+
+func TestNewJSONClient(t *testing.T) {
+	var testData = []struct {
+		client *http.Client
+		pubKey string
+		errstr string
+	}{
+		{nil, "bogus", "no PEM block"},
+		{nil, testdata.RsaPublicKeyPEM, ""},
+		{nil, testdata.EcdsaPublicKeyPEM, ""},
+		{nil, testdata.DsaPublicKeyPEM, "Unsupported public key type"},
+		{nil, testdata.RsaPublicKeyPEM + "bogus", "extra data found"},
+	}
+	for i, tc := range testData {
+		client, err := New("http://127.0.0.1", tc.client, tc.pubKey)
+		if tc.errstr == "" {
+			if err != nil {
+				t.Errorf("#%d New() failed: %s", i, err.Error())
```

Comment:
I've had the advice to include as the first field of your test struct a descriptive string:

```go
tests := []struct {
  desc string
  other stuff
}{
  {"key with no PEM block", "..", ".."}
  ...
}
```
Then in error messages you'd make the first thing you print the test.desc rather than the index in the test table.
The idea being that you can more easily eyeball test output and see what the purpose of the test was when it fails.